logo.js

```diff
@@ -1673,23 +1837,31 @@ function LogoInterpreter(turtle, stream, savehook)
   // 6.3 Turtle and Window Control
   //
 
-  def(["showturtle", "st"], function() { turtle.showturtle(); });
-  def(["hideturtle", "ht"], function() { turtle.hideturtle(); });
-  def("clean", function() { turtle.clear(); });
-  def(["clearscreen", "cs"], function() { turtle.clearscreen(); });
+  def(["showturtle", "st"], function() { return turtle.showturtle(); });
+  def(["hideturtle", "ht"], function() { return turtle.hideturtle(); });
+  def("clean", function() { return turtle.clear(); });
+  def(["clearscreen", "cs"], function() { return turtle.clearscreen(); });
 
-  def("wrap", function() { turtle.setturtlemode('wrap'); });
-  def("window", function() { turtle.setturtlemode('window'); });
-  def("fence", function() { turtle.setturtlemode('fence'); });
+  def("wrap", function() { return turtle.setturtlemode('wrap'); });
+  def("window", function() { return turtle.setturtlemode('window'); });
+  def("fence", function() { return turtle.setturtlemode('fence'); });
 
-  def("fill", function() { turtle.fill(); });
+  def("fill", function() { return turtle.fill(); });
 
   def("filled", function(fillcolor, statements) {
     fillcolor = sexpr(fillcolor);
     statements = reparse(lexpr(statements));
     turtle.beginpath();
     try {
-      self.execute(statements);
+      return self.execute(statements).then(
+        function (result) {
+          turtle.fillpath(fillcolor);
+          return result;
+        },
+        function (err) {
+          turtle.fillpath(fillcolor);
```

Comment:
Can we add a comment that this is to handle exceptions as loop/proc terminators? (I know, I wrote the try/finally originally, but surprised me when re-reading the code).